utils/rr_graph_walk/rr_graph_walk.py

```diff
@@ -0,0 +1,394 @@
+#!/usr/bin/env python3
+"""
+This utility script allows to walk through the routing graph from a given
+starting node id to a given target node id. If the target node id is not
+given then it lists all available routes which start at the starting node.
+"""
+
+import sys
+import argparse
+import gc
+from collections import namedtuple
+from copy import deepcopy
+
+import lxml.etree as ET
+import lxml.objectify as objectify
+
+from progressbar import progressbar
+
+# =============================================================================
+
+
+class RoutingGraph(object):
+
+    Node = namedtuple("Node", "id type xlow ylow xhigh yhigh")
+    WalkContext = namedtuple("WalkContext", "node_id depth")
+
+    def __init__(self, xml_file):
+
+        # Load and parse the XML
+        print("Loading XML file...")
+        parser = ET.XMLParser(remove_comments=True)
+        xml_tree = objectify.parse(xml_file, parser=parser)
+        xml_root = xml_tree.getroot()
+
+        self.nodes = {}
+
+        self.edges_from = {}
+        self.edges_to = {}
+
+        # Import nodes
+        self._import_nodes(xml_root)
+        # Import edges
+        self._import_edges(xml_root)
+
+        del xml_root
+        del xml_tree
+        del parser
+        gc.collect()
+
+    def _import_nodes(self, xml_root):
```

Comment:
Consider using a streaming parser to avoid the large memory cost?  Might even be faster?

Reply:
https://docs.python.org/2/library/xml.etree.elementtree.html#xml.etree.ElementTree.iterparse

Reply:
@litghost That was something I planned but didn't have time for. Now it is done. The runtime is increased by about 2min but the memory usage is dramatically reduced:

```
Non-iterative XML parsing:
	Elapsed (wall clock) time (h:mm:ss or m:ss): 13:30.93
	Maximum resident set size (kbytes): 17543404

Iterative XML parsing:
	Elapsed (wall clock) time (h:mm:ss or m:ss): 15:15.95
	Maximum resident set size (kbytes): 2244860
```

Reply:
@mkurc-ant Is that 15 minutes runtime?

Reply:
Yes it is. I implemented the depth first search algorithm which performs the full search over the rr graph. I also have a C++ implementation which I did expecting performance issues with Python. And the C++ version is about 2x faster.

Reply:
Routers are typically breadth first search (https://en.wikipedia.org/wiki/Maze_solving_algorithm#Shortest_path_algorithm) not depth first.

Reply:
But still I wanted this tool also to be able to find all possible routes from/to a node. And I found it easier to implement the tracking of the route using depth first search.

Breadth first search would explore broader area of the graph faster but I don't think it would hit a specific target node any faster.